proxy/InkAPI.cc

```diff
@@ -5913,33 +5922,34 @@ TSHttpTxnReenable(TSHttpTxn txnp, TSEvent event)
 }
 
 TSReturnCode
-TSHttpArgIndexReserve(const char *name, const char *description, int *arg_idx)
+TSHttpArgIndexReserve(UserArg::Type type, const char *name, const char *description, int *ptr_idx)
 {
-  sdk_assert(sdk_sanity_check_null_ptr(arg_idx) == TS_SUCCESS);
+  sdk_assert(sdk_sanity_check_null_ptr(ptr_idx) == TS_SUCCESS);
+  sdk_assert(0 <= type && type < UserArg::Type::COUNT);
 
-  int ix = ink_atomic_increment(&next_argv_index, 1);
+  int idx = UserArgIdx[type]++;
 
-  if (ix < HTTP_SSN_TXN_MAX_USER_ARG) {
-    state_arg_table[ix].name     = ats_strdup(name);
-    state_arg_table[ix].name_len = strlen(state_arg_table[ix].name);
-    if (description) {
-      state_arg_table[ix].description = ats_strdup(description);
-    }
-    *arg_idx = ix;
+  if (idx < TS_HTTP_MAX_USER_ARG) {
+    UserArg &arg(UserArgTable[type][idx]);
+    arg.name        = name ? name : "*UNKNOWN*";
```

Comment:
Do we really want to handle the case of name being a null pointer? I think it'd be better to do an sdk_sanity_check_null_ptr on it, and document that name (and perhaps also description?) can not be NULL or an empty string?

Other than that, loving this.

Reply:
I think we can let `description` slide, but looking ahead to [plugin coordination](https://solidwallofcode.github.io/plugin-coordination.en.html) we definitely want to require a valid name.